src/sagas/tokens.js

```diff
@@ -0,0 +1,291 @@
+  }
+}
+
+/**
+ * This saga will consume the fetchTokenBalanceChannel for TOKEN_FETCH_BALANCE_REQUEST actions
+ * and wait until the TOKEN_FETCH_BALANCE_SUCCESS action is dispatched with the specific tokenId
+ */
+function* fetchTokenMetadataConsumer(fetchTokenMetadataChannel) {
+  while (true) {
+    const action = yield take(fetchTokenMetadataChannel);
+
+    yield fork(fetchTokenMetadata, action);
+
+    // Wait until the success action is dispatched before consuming another action
+    yield take(
+      specificTypeAndPayload([
+        types.TOKEN_FETCH_METADATA_SUCCESS,
+        types.TOKEN_FETCH_METADATA_FAILED,
+      ], {
+        tokenId: action.tokenId,
+      }),
+    );
+  }
+}
+
+/**
+ * Fetch a single token from the metadataApi
+ *
+ * @param {String} action.tokenId The token to fetch from the metadata api
+ */
+export function* fetchTokenMetadata({ tokenId }) {
+  const { network } = yield select((state) => state.serverInfo);
+
+  try {
+    // Retry mechanism
+    for (let i = 0; i <= METADATA_MAX_RETRIES; i += 1) {
+      try {
+        const data = yield call(metadataApi.getDagMetadata, tokenId, network);
+
+        yield put({
+          type: types.TOKEN_FETCH_METADATA_SUCCESS,
+          tokenId,
+          data: get(data, tokenId, null),
+        });
+        return;
+      } catch (e) {
+        yield delay(1000); // Wait 1s before trying again
+      }
+    }
+
+    throw new Error(`Max retries requesting metadata for ${tokenId}`);
+  } catch (e) {
+    yield put({
+      type: types.TOKEN_FETCH_METADATA_FAILED,
+      tokenId,
+    });
+    // eslint-disable-next-line
+    console.log('Error downloading metadata of token', tokenId);
+  }
+}
+
+function* fetchTokenHistory(action) {
+  const { tokenId, force } = action;
+
+  try {
+    const wallet = yield select((state) => state.wallet);
+    const tokenHistory = yield select((state) => get(state.tokensHistory, tokenId));
+
+    if (!force && tokenHistory && tokenHistory.oldStatus === TOKEN_DOWNLOAD_STATUS.READY) {
+      // The data is already loaded, we should dispatch success
+      yield put(tokenFetchHistorySuccess(tokenId, tokenHistory.data));
+      return;
+    }
+
+    const response = yield call(wallet.getTxHistory.bind(wallet), { token_id: tokenId });
+    const data = response.map((txHistory) => helpers.mapTokenHistory(txHistory, tokenId));
+
+    yield put(tokenFetchHistorySuccess(tokenId, data));
+  } catch (e) {
+    yield put(tokenFetchHistoryFailed(tokenId));
+  }
+}
+
+/**
+ * This saga will route the actions dispatched from SET_TOKEN and NEW_TOKEN to the
+ * TOKEN_FETCH_BALANCE_REQUESTED saga, the idea is to load the balance for new tokens
+ * registered or created on the app.
+ */
+function* routeTokenChange(action) {
+  const wallet = yield select((state) => state.wallet);
+
+  if (!wallet || !wallet.isReady()) {
+    return;
+  }
+
+  switch (action.type) {
+    default:
+    case 'new_tokens':
+      for (const token of action.payload.tokens) {
```

Comment:
This `tokens` array will have all the tokens in the storage, not only the new ones. Is that expected?

Reply:
It is hard at this point to identify what tokens are new since the changes were already commited to the redux store

It shouldn't be a problem to dispatch though, as the tokens that already exist won't be downloaded again (as they are already `ready`)